Threading logic: merge sbarratt/threading

sbarratt opened the pull request:
Since SCS releases the GIL, we can simply wrap the diffcp function with multi-threading .

akshayka commented on diffcp/cone_program.py:
```diff
@@ -28,6 +30,11 @@ def dpi(z, cones):
+def solve_and_derivative_batch(As, bs, cs, cone_dicts, n_jobs=4, warm_starts=None, **kwargs):
+    batch_size = len(As)
+    return Parallel(n_jobs=n_jobs, backend="threading")(
+        delayed(solve_and_derivative)(As[i], bs[i], cs[i], cone_dicts[i], warm_start=None if warm_starts is None else warm_starts[i], **kwargs) for i in (range(batch_size)))
```
nit, lines shouldn't exceed 80 columns

akshayka commented on tests.py:
```diff
@@ -279,6 +280,36 @@ def test_warm_start(self):
+        toc = time.time()
+        parallel_time = toc - tic
+
+        print(parallel_time, serial_time)
```
nit, if you want to print these stats, include some text (e.g., "parallel time {:f}, serial time {:f}".format(parallel_time, serial_time))

sbarratt commented:
Ok, I've incorporated these changes. I've also added the newest version of scs to the setup.py requirements. This stops the AA warning messages being printed. (https://github.com/cvxgrp/scs/issues/115)

akshayka commented:
Thanks!
